#imbalanced dataset
a datset with unequal class distribution
A dataset is imbalanced when the number of examples in each class is very different.

Example:
In diabetes prediction:

Class (Outcome)	Count
0 (Non-diabetic)	500
1 (Diabetic)	100

🔴 The model may learn to always predict "0" because it gives 83% accuracy by doing nothing smart!

In [3]:
import numpy as np
import pandas as pd


In [4]:
credit_card_data=pd.read_csv('credit_data.csv')

In [5]:
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

0--> legit 
1--> fraud transaction

In [6]:
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [7]:
# this data doesnot contain any missing value
# determine the distribution of two classes
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [9]:
credit_card_data.shape

(284807, 31)

In [ ]:
# it is an imbalanced datset as for 0 there are so much data nd for 1 its less

In [11]:
# separating the legit and fraudulent transactions
legit=credit_card_data[credit_card_data.Class==0]# it will contain all the rows which contain class value as 0

fraud=credit_card_data[credit_card_data.Class==1]# it will contain all the rows which contain class value as 1


In [13]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


#undersampling--buliding a sample dataset contianing similar distribution of legit and fraudulent transaction
number of farudulent transactions--> 492

In [14]:
legit_sample=legit.sample(n=492)

In [18]:
print(legit_sample.shape)

(492, 31)


In [19]:
# concatenate the 2 dataframes
new_dataset=pd.concat([legit_sample,fraud],axis=0) # we want to conactenate row wise so axis=0


In [20]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
60557   49381.0 -7.473506  6.635679 -5.316242  1.321083 -4.406693 -1.563825   
71579   54348.0 -4.188037  3.379227 -0.334106 -1.946685 -1.446410 -0.725538   
115860  74047.0 -3.080834 -1.495782  1.605817  2.063578 -1.323864  1.149444   
78167   57399.0 -0.805551  1.822450 -0.505021  0.386030  0.757969 -0.265001   
26462   34079.0 -0.644663  0.503441 -0.087584 -0.746683  1.771460  4.017483   

              V7        V8        V9  ...       V21       V22       V23  \
60557  -4.629767  5.990822 -2.026305  ...  0.812353  0.171201  0.670319   
71579  -0.568926  1.449800  1.684028  ... -0.302829 -0.193376  0.216273   
115860  0.190644  0.671708  0.055288  ...  0.220945  0.924422 -0.699777   
78167   0.990222 -0.172464  0.796244  ... -0.103016  0.391949 -0.027232   
26462  -0.544233  1.493098  0.109230  ... -0.012846 -0.043123 -0.067104   

             V24       V25       V26       V27       V28  Amount  Class  
60557   0.262322  0.280976 -0.372574 -2.649876 -0.537537    1.00      0  
71579   0.186190  0.082652  0.695122  0.421567 -0.089549    0.75      0  
115860  0.391717  0.335123  0.067861  0.658328 -0.539509  424.00      0  
78167  -0.803363 -0.481406 -0.336041  0.807431  0.293798    1.29      0  
26462   1.057902 -0.171423  0.326207  0.017721  0.088770   29.90      0  

[5 rows x 31 columns]

In [21]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [22]:
new_dataset.shape

(984, 31)

In [23]:
new_dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

#in undersampling we might loose important data
so other methods are to be used

🔝 1. Class Weighting (class_weight='balanced')
🔥 Most common for Logistic Regression, SVM, Random Forest

Tells model: “Pay more attention to the minority class”

✅ Why it’s popular:
Very easy to apply

No data is changed

Built into most classifiers (no need to install extra libraries)

🔍 Example:

from sklearn.linear_model import LogisticRegression
model = LogisticRegression(class_weight='balanced')
Also works with:

RandomForestClassifier(class_weight='balanced')

SVC(class_weight='balanced')

DecisionTreeClassifier(class_weight='balanced')

🔝 2. SMOTE (Synthetic Minority Over-sampling Technique)
🔥 Most effective when data is very imbalanced

Creates synthetic (fake but realistic) data points for the minority class

✅ Why it's important:
Doesn’t lose data (like undersampling)

Gives balanced training data

Works well with many ML models

🔍 Example:

from imblearn.over_sampling import SMOTE

smote = SMOTE()
X_resampled, y_resampled = smote.fit_resample(X, y)
Use after data preprocessing, before model training.

🎯 When to Use What?
Situation	Best Method
Model supports class_weight	✅ Use class_weight='balanced'
Dataset is very imbalanced	✅ Use SMOTE
Dataset is small	        ✅ Use SMOTE
Dataset is large	        ❌ Avoid undersampling (lose info)
Using tree models(Randomfore ✅ Use class_weight or SMOTE

✅ Final Recommendation (most used combo):
For your diabetes or disease prediction project:


from imblearn.over_sampling import SMOTE
from sklearn.linear_model import LogisticRegression

#Step 1: Apply SMOTE
sm = SMOTE()
X_res, y_res = sm.fit_resample(X, y)

#Step 2: Train with class weight
model = LogisticRegression(class_weight='balanced')
model.fit(X_res, y_res)
Let me know if you'd like a full notebook example using SMOTE and class weights with your diabetes dataset!

What Does SMOTE Actually Do?
SMOTE doesn't copy existing data — it creates new, synthetic (fake but realistic) data points for the minority class by interpolating between existing ones.

✅ SMOTE’s Core Idea:
For each minority class sample, SMOTE:

Finds its k-nearest neighbors (default k=5)

Randomly picks one of those neighbors

Creates a new point somewhere between them

This is done in feature space, not in label space.

🔢 Let’s Take a Numerical Example
Let’s say you have 2 features:

👇 Original Minority Samples:
Sample	Feature 1 (f1)	Feature 2 (f2)
A	2.0	3.0
B	4.0	5.0

Now, SMOTE will:

Pick Sample A

Find that Sample B is a neighbor

Create a new sample between A and B

✏️ Formula Used:
For a feature:

ini
Copy
Edit
synthetic_feature = A + (B - A) * random_number_between_0_and_1
Let’s say:

Random number = 0.5

Then for both features:

yaml
Copy
Edit
Feature 1: 2 + (4 - 2) * 0.5 = 3.0
Feature 2: 3 + (5 - 3) * 0.5 = 4.0
✅ So the new synthetic point will be:
[3.0, 4.0] — which lies halfway between A and B.

🧠 What if random number = 0.8?
yaml
Copy
Edit
Feature 1: 2 + (4 - 2) * 0.8 = 3.6
Feature 2: 3 + (5 - 3) * 0.8 = 4.6
So, new point = [3.6, 4.6]